python/cuml/svm/linear.pyx

```diff
@@ -0,0 +1,650 @@
+           signature - i.e. how it appears in inspect/help/docs.
+           __init__ method must have `**kwargs` argument for this to
+           make sense.
+
+        2. Applies string.format() to the class docstring with the globals()
+           in the scope of the __init__ method.
+           This allows to write variable names (e.g. some constants) in docs,
+           such that they are substituted with their actual values.
+           Useful for reexporting default values from somewhere else.
+    '''
+
+    def __new__(cls, name, bases, attrs):
+
+        def get_class_params(init, parents):
+            # collect the keyword arguments from the class hierarchy
+            params = OrderedDict()
+            for k in parents:
+                params.update(
+                    get_class_params(getattr(k, '__init__', None), k.__bases__)
+                )
+            if init is not None:
+                sig = inspect.signature(init)
+                for k, v in sig.parameters.items():
+                    if v.kind == inspect.Parameter.KEYWORD_ONLY:
+                        params[k] = v
+                del sig
+            return params
+
+        init = attrs.get('__init__', None)
+        if init is not None:
+            # insert keyword arguments from parents
+            ppams = get_class_params(init, bases)
+            sig = inspect.signature(init)
+            params = [
+                p for p in sig.parameters.values()
+                if p.kind != inspect.Parameter.KEYWORD_ONLY
+            ]
+            params[1:1] = ppams.values()
+            attrs['__init__'].__signature__ = sig.replace(parameters=params)
+            del sig
+
+            # format documentation -- replace variables with values
+            doc = attrs.get('__doc__', None)
+            if doc is not None:
+                globs = init.__globals__.copy()
+                globs[name] = type(name, (), attrs)
+                attrs['__doc__'] = \
+                    re.sub(r"\{ *([^ ]+) *\}", r"{\1}", doc).format(**globs)
+                del globs
+            del doc
+        del init
+        return super().__new__(cls, name, bases, attrs)
+
+
+class LinearSVM(Base, metaclass=WithReexportedParams):
+
+    _model_: typing.Optional[LinearSVMWrapper]
+
+    coef_ = CumlArrayDescriptor()
+    intercept_ = CumlArrayDescriptor()
+    classes_ = CumlArrayDescriptor()
+    probScale_ = CumlArrayDescriptor()
+
+    @property
+    def model_(self) -> LinearSVMWrapper:
+        if self._model_ is None:
+            raise AttributeError(
+                'The model is not trained yet (call fit() first).')
+        return self._model_
+
+    def __getstate__(self):
+        state = self.__dict__.copy()
+        state['_model_'] = None
+        return state
+
+    def __init__(self, **kwargs):
+        # `tol` is special in that it's not present in get_param_names,
+        # so having a special logic here does not affect pickling/cloning.
```

Comment:
Why is it not present in get_param_names?

Reply:
`tol` is not a parameter of the underlying model (we have `grad_tol` and `change_tol` instead). So I don't store it anywhere - keep it only for compatibility with sklearn. I see it as "syntactic sugar" for the `LinearSV*` constructors :)